tlv/stream.go

```diff
@@ -0,0 +1,259 @@
+func (t ErrUnknownRequiredType) Error() string {
+	return fmt.Sprintf("unknown required type: %d", t)
+}
+
+// RecordEncoder specifies a codec for serializing a given record.
+type RecordEncoder func(io.Writer, interface{}) error
+
+// RecordDecoder specifies a codec for deserializing given record. The final
+// argument is the expected length, and the method should return an error if
+// thee number of bytes read does not match this value.
+type RecordDecoder func(io.Reader, interface{}, uint64) error
+
+// Stream defines a TLV stream that can be used for encoding or decoding a set
+// of TLV Records.
+type Stream struct {
+	records []Record
+	buf     [8]byte
+}
+
+// NewStream creates a new TLV Stream given an encoding codec, a decoding codec,
+// and a set of known records.
+func NewStream(records ...Record) *Stream {
+	// Assert that the ordering of the Records is canonical and appear in
+	// ascending order of type.
+	var min Type
+	var overflow bool
+	for _, record := range records {
+		if overflow || record.typ < min {
+			panic(ErrStreamNotCanonical.Error())
+		}
+		if record.encoder == nil {
+			panic(ErrNoEncoder.Error())
+		}
+		if record.decoder == nil {
+			panic(ErrNoDecoder.Error())
+		}
+		if record.typ == math.MaxUint64 {
+			overflow = true
+		}
+		min = record.typ + 1
+	}
+
+	return &Stream{
+		records: records,
+	}
+}
+
+// Encode writes a Stream to the passed io.Writer. Each of the Records known to
+// the Stream is written in ascending order of their type so as to be canonical.
+//
+// The stream is constructed by concatenating the individual, serialized Records
+// where each record has the following format:
+//    [varint: type]
+//    [varint: length]
+//    [length: value]
+//
+// An error is returned if the io.Writer fails to accept bytes from the
+// encoding, and nothing else. The ordering of the Records is asserted upon the
+// creation of a Stream, and thus the output will be by definition canonical.
+func (s *Stream) Encode(w io.Writer) error {
+	// Iterate through all known records, if any, serializing each record's
+	// type, length and value.
+	for i := range s.records {
+		rec := &s.records[i]
+
+		// Write the record's type as a varint.
+		err := WriteVarInt(w, uint64(rec.typ), &s.buf)
+		if err != nil {
+			return err
+		}
+
+		// Write the record's length as a varint.
+		err = WriteVarInt(w, rec.Size(), &s.buf)
+		if err != nil {
+			return err
+		}
+
+		// Encode the current record's value using the stream's codec.
+		err = rec.encoder(w, rec.value, &s.buf)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// Decode deserializes TLV Stream from the passed io.Reader. The Stream will
+// inspect each record that is parsed and check to see if it has a corresponding
+// Record to facilitate deserialization of that field. If the record is unknown,
+// the Stream will discard the record's bytes and proceed to the subsequent
+// record.
```

Comment:
> If the record is unknown,
> // the Stream will discard the record's bytes and proceed to the subsequent
> // record.

Is this subject to the even/odd rule? Also perhaps we should allow a "strict" stream that'll always error our rather than silently go to the next record to be decoded.

Reply:
yes a strict mode would definitely be useful. the original version of this pr included that, but i removed it for now to simplify the diff